libs/src/api/user.js

```diff
@@ -683,6 +684,47 @@ class Users extends Base {
   _cleanUserMetadata (metadata) {
     return pick(metadata, USER_PROPS.concat('user_id'))
   }
+
+  async _updateReplicaSet (userId, metadata) {
+    // Attempt to update through UserReplicaSetManagerClient if present
+    if (!this.contracts.UserReplicaSetManagerClient) {
+      await this.contracts.initUserReplicaSetManagerClient()
+    }
+
+    // If still uninitialized, proceed with legacy update - else move forward with new contract update
+    if (!this.contracts.UserReplicaSetManagerClient) {
+      const { txReceipt: updateEndpointTxReceipt } = await this.contracts.UserFactoryClient.updateCreatorNodeEndpoint(
+        userId,
+        metadata.creator_node_endpoint
+      )
+      return updateEndpointTxReceipt
+    }
+
+    let primaryEndpoint = CreatorNode.getPrimary(metadata.creator_node_endpoint)
+    let secondaries = CreatorNode.getSecondaries(metadata.creator_node_endpoint)
+    let primarySpID = await this._retrieveSpIDFromEndpoint(primaryEndpoint)
+    let secondary1SpID = await this._retrieveSpIDFromEndpoint(secondaries[0])
+    let secondary2SpID = await this._retrieveSpIDFromEndpoint(secondaries[1])
+    // Update in new contract
+    let tx = await this.contracts.UserReplicaSetManagerClient.updateReplicaSet(
+      userId,
+      primarySpID,
+      [secondary1SpID, secondary2SpID]
+    )
+    return tx
+  }
+
+  async _retrieveSpIDFromEndpoint (endpoint) {
+    let cachedSpID = await getSpIDFromEndpoint(endpoint)
+    let spID = cachedSpID
+    if (!spID) {
+      let spEndpointInfo = await this.ethContracts.ServiceProviderFactoryClient.getServiceProviderInfoFromEndpoint(
+        endpoint
+      )
+      spID = spEndpointInfo.spID
+    }
+    return spID
+  }
```

Comment:
as is, a single failed web3 call in this function will short circuit the full _updateReplicaSet() call - this seems correct,just confirming also we should explicitly comment. could also think about retries but would be down to just throw

also if `spEndpointInfo` on line 721 is null/empty then spID on line 724 will also be empty, at that point i thnk we should error instead of writing this to chain

Reply:
added comment explaining error